File: LinkedList/05.ReverseNodesKGroup/ReverseNodesKGroup.js
// 以 k 为一组翻转链表
// https://leetcode.com/problems/reverse-nodes-in-k-group/
//
/**
 * 题解，递归解法
 * 1. 尝试将头结点后移动 k 位，如果不足 k 位，直接返回头结点
 * 2. 如果存在 k 个结点，用新的头结点，递归
 * 3. 得到的结果连接到头 k 的节点的逆序链表后面
 * 4. 重新设置头结点，返回
 */
var reverseKGroup = function(head, k) {
	let curr = head;
	let count = 0;
	while (curr && count != k) {
		curr = curr.next;
		++count;
	}
	if (count === k) {
		curr = reverseKGroup(curr, k);

		while (count-- > 0) {
			let tmp = head.next;
			head.next = curr;
			curr = head;
			head = tmp;
		}
		head = curr;
	}
	return head;
};

/**
 * 迭代解法
 * 遍历链表，如果有 k 个，就翻转，然后将翻转后的 group 与已翻转和未翻转的链表合并
 */

 // Runtime: 68 ms, faster than 92.95% of JavaScript online submissions for Reverse Nodes in k-Group.
var reverseKGroup2 = function(head, k) {
	// 初始化一个节点用于指向头结点
	var dump = jump = new ListNode(0);
	dump.next = head;

	// l 前一个 group 的 head
	// r 后一个 group 的 head
	var l = r = head;

	while (true) {
		var count = 0;
		// r 移到下一组 group 的头部
		while (r && count < k) {
			r = r.next;
			count++;
		}
		// 翻转 [l, r) 之间的链表
		if (count === k) {
			let cur = l, pre = r;
			while(count--) {
				let tmp = cur.next;
				cur.next = pre;
				pre = cur;
				cur = tmp;
			}
			// 翻转后 pre 为当前 group 头节点
			// jump 为前一个 group 的尾结点或辅助节点
			// 将前一个 group 的尾结点指向当前 group 的节点
			jump.next = pre;
			// jump 指向当前 group 的最后一个节点
			jump = l;
			// l 指向下一组 group 的头节点
			l = r;
		} else {
			// 因为 dump 等于 jump 的初始
			// 而 jump 的初值的 next 指向链表的头部
			return dump.next;
		}
	}
};

File: LinkedList/05.ReverseNodesKGroup/ReverseNodesKGroup.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* _reverseKGroup(ListNode* head) {
        ListNode *prev = nullptr, *curr = head, *back;
        while (curr) {
            back = curr->next;
            curr->next = prev;
            prev = curr;
            curr = back;
        }
        return prev;
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode *dummy = new ListNode(0); dummy->next = head;
        ListNode *end, *begin = head; // 链表首尾
        ListNode *back, *prev = dummy; // 前驱后继
        while(1){
            end = begin;
            for(int i=1; i<k && end; ++i) { end = end->next; }
            if(end == nullptr) { break; }
            back = end->next; // 记录后继
            // 链表反转
            end->next = nullptr;
            prev->next = _reverseKGroup(begin); // 连接前面已翻转链表
            begin->next = back; // 连接后面未翻转链表
            prev = begin; // 更新前驱
            begin = back; // 更新链表首
        }
        return dummy->next;
    }
};